### PR TITLE
Typescript handoff and Java side changes

### PR DESCRIPTION
**Pull request by rmohan20**
1. PmdCatalogWrapper combines inbuilt PMD rules and user's custom rules
2. PmdCatalogWrapper also builds java parameters, which contain language to path mapping
3. Java Main handles the new parameters
4. Merging java changes from rmohan/customRules branch to handle all types of resources as one (Jar, library of Jars, directory with compiled class files)

### CODE REVIEW

File: src/lib/pmd/PmdCatalogWrapper.ts [hunk 1]
```diff
@@ -84,17 +84,64 @@ export class PmdCatalogWrapper extends PmdSupport {
...
+  protected async buildClasspath(): Promise<string[]> {
     // TODO: This probably isn't where the JAR ought to live. Once the JAR's home is finalized, come back to this.
     const catalogerLibs = `${PMD_CATALOGER_LIB}/*`;
-    return super.buildClasspath().concat([catalogerLibs]);
+    const classpathEntries = await super.buildClasspath();
+    classpathEntries.push(catalogerLibs);
+    return classpathEntries;
+  }
+
+  private async buildCatalogerParameters(): Promise<string[]> {
```
Comment (rmohan20) [inline src/lib/pmd/PmdCatalogWrapper.ts]: Crux of the change is here. This portion of the code fetches the custom rule path entries and merges PMD's inbuilt rules into the same Map. It then builds each argument as language=/rule/path,/rule/path2

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/Main.java [hunk 2]
```diff
@@ -1,25 +1,65 @@
...
+import java.util.stream.Collectors;
 
 import sfdc.sfdx.scanner.ExitCode;
 
 public class Main {
+
+  private static String DIVIDER = "=";
+  private static String COMMA = ",";
+  /**
+   * MAin entry to PMD Rule Catalog builder
```
Comment (rmohan20) [inline pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/Main.java]: weird caps 🤦‍♀

File: src/lib/CustomRulePathManager.ts [hunk 3]
```diff
@@ -64,6 +64,11 @@ export class CustomRulePathManager {
     return await this.saveCustomClasspaths();
   }
 
+  public async getRulePathEntries(engine: ENGINE): Promise<Map<string, Set<string>>> {
```
Comment (jfeingold35) [inline src/lib/CustomRulePathManager.ts]: Nitpick: `getRulePathEntriesForEngine()`?

File: src/lib/pmd/PmdCatalogWrapper.ts [hunk 4]
```diff
@@ -84,17 +84,64 @@ export class PmdCatalogWrapper extends PmdSupport {
...
   }
 
-  protected buildCommandArray(): [string, string[]] {
+  protected async buildCommandArray(): Promise<[string, string[]]> {
     const command = 'java';
     // NOTE: If we were going to run this command from the CLI directly, then we'd wrap the classpath in quotes, but this
     // is intended for child_process.spawn(), which freaks out if you do that.
-    const args = ['-cp', this.buildClasspath().join(':'), MAIN_CLASS, PMD_LIB, PMD_VERSION, SUPPORTED_LANGUAGES.join(',')];
+    const classpathEntries = await this.buildClasspath();
+    const parameters = await this.buildCatalogerParameters();
```
Comment (jfeingold35) [inline src/lib/pmd/PmdCatalogWrapper.ts]: Performance: These two lines of code could be executed at the same time if you replace them with `const [classpathEntries, parameters] = await Promise.all([this.buildClasspath(), this.buildCatalogerParameters()]);`.

File: src/lib/pmd/PmdCatalogWrapper.ts [hunk 5]
```diff
@@ -84,17 +84,64 @@ export class PmdCatalogWrapper extends PmdSupport {
...
+    const args = ['-cp', classpathEntries.join(':'), MAIN_CLASS, ...parameters];
     return [command, args];
   }
 
-  protected buildClasspath(): string[] {
+  protected async buildClasspath(): Promise<string[]> {
     // TODO: This probably isn't where the JAR ought to live. Once the JAR's home is finalized, come back to this.
```
Comment (jfeingold35) [inline src/lib/pmd/PmdCatalogWrapper.ts]: Delete this TODO. I'm pretty sure we've settled on where the JAR lives.

File: src/lib/pmd/PmdCatalogWrapper.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ import {AnyJson} from '@salesforce/ts-types';
 import fs = require('fs');
 import {Rule} from '../../types';
 import {RuleFilter, RULE_FILTER_TYPE} from '../RuleManager';
-import {PmdSupport, PMD_LIB, PMD_VERSION} from './PmdSupport';
```
Comment (jfeingold35) [inline src/lib/pmd/PmdCatalogWrapper.ts]: Don't you need to add PMD_LIB to the start of the PMD standard JAR paths?

File: src/lib/pmd/PmdCatalogWrapper.ts [hunk 7]
```diff
@@ -84,17 +84,64 @@ export class PmdCatalogWrapper extends PmdSupport {
...
+      rulePathEntries.get(language).add(pmdJarName);
+    });
+  }
+
+
+  /**
+   * PMD library holds the same naming structure for each language
+   */
+  private derivePmdJarName(language: string): string {
+    return "pmd-" + language + "-" + PMD_VERSION + ".jar";
```
Comment (jfeingold35) [inline src/lib/pmd/PmdCatalogWrapper.ts]: This needs to start with the PMD_LIB path, unless I'm mistaken.

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java [hunk 8]
```diff
@@ -13,12 +12,14 @@
...
-  private String pmdVersion;
-  private String pmdPath;
-  private List<String> languages;
-  // These two maps are how we know which files we need to scan for each language.
-  private Map<String,List<String>> categoryPathsByLanguage = new HashMap<>();
-  private Map<String,List<String>> rulesetPathsByLanguage = new HashMap<>();
+//  private String pmdVersion;
+//  private String pmdPath;
+//  private String customClasspathMapping;
+//  private List<String> languages;
```
Comment (jfeingold35) [inline pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java]: Why are these still here?

File: src/lib/pmd/PmdCatalogWrapper.ts (hunk 4 above)
Comment (rmohan20) [reply src/lib/pmd/PmdCatalogWrapper.ts]: Good to know! Thanks.

File: src/lib/pmd/PmdCatalogWrapper.ts (hunk 6 above)
Comment (rmohan20) [reply src/lib/pmd/PmdCatalogWrapper.ts]: Yup, burnt my finger. Fixed it.

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java (hunk 8 above)
Comment (rmohan20) [reply pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java]: Good catch. Cleaned up.

File: src/lib/pmd/PmdCatalogWrapper.ts (hunk 7 above)
Comment (rmohan20) [reply src/lib/pmd/PmdCatalogWrapper.ts]: You are right.

File: src/lib/CustomRulePathManager.ts (hunk 3 above)
Comment (rmohan20) [reply src/lib/CustomRulePathManager.ts]: This is an API convention where Engine is implied when we pass it as parameter.

**Pull request merged**